Implement list of custom electrum servers

misovan opened the pull request:
This PR implements the following features:

- add or remove custome electrum server to the new list managed in Netwrok > Electrum server settings
- select server from the list to be connected to (after restart)
- remove server from the list
- delete all custom servers and return to default bluewallet servers

Rational is that users might want to connect to their electrum server while using wallet in their private network but be able to quickly switch to another server when on public network (might still be their own but exposed e.g. through ssh tunnel to their device or public one they trust)

marcosrdz commented:
Im getting a looped Error in connect() function when I click Save
![Simulator Screen Shot - iPhone 11 - 2020-09-04 at 11 56 01](https://user-images.githubusercontent.com/4793122/92260248-9c72b880-eea5-11ea-9f58-53ee546708a8.png)

misovan commented:
Thanks fo giving a try. I will try to simulate it on fresh install (did not happen to me and I believe I did not touch connect function)

misovan commented:
OK, I was able to simulate the error that occured when incorrect host or port were submitted. In such case, BlueElectrum.testConnection was periodically trying to reconnect to wrong address. I fixed the bug in the function and on top I slightly revamped the original form to add new electrum server in electrum settings - it now allows selecting either SSL (TLS) or unencrypted port and it got some basic validations.

![Screenshot_20200905-001335_BlueWallet](https://user-images.githubusercontent.com/1290973/92288525-797bef00-ef0d-11ea-89f5-6706623e0751.png)

marcosrdz commented:
@misovan do you mind granting me write access to your branch so I can make UI modifications?

misovan commented:
Hi @marcosrdz may I kindly ask you for quick guidance regarding remaining DeepScan check? It reports multiple issues in files I did not change in this PR and then one in my code I am not sure how to efficiently address: 
https://deepscan.io/dashboard/#view=project&pid=6979&iid=39254385&tid=5209&subview=pull-request&prid=471778

Thank you upfront for your help.

marcosrdz commented:
> Hi @marcosrdz may I kindly ask you for quick guidance regarding remaining DeepScan check? It reports multiple issues in files I did not change in this PR and then one in my code I am not sure how to efficiently address: 
> 
> https://deepscan.io/dashboard/#view=project&pid=6979&iid=39254385&tid=5209&subview=pull-request&prid=471778
> 
> 
> 
> Thank you upfront for your help.

Hello! I'll be working on updating this branch really soon. I'll address the deepscan result.

Overtorment commented:
fyi sometimes deepscan is barking at the wrong tree and is safe to ignore.
Ill be review this later.

Overtorment commented on blue_modules/BlueElectrum.js:
```diff
@@ -525,9 +541,10 @@ module.exports.testConnection = async function (host, tcpPort, sslPort) {
     await client.server_version('2.7.11', '1.4');
     await client.server_ping();
-    client.close();
+    await client.close();
```
anyway I think this part was refactored massively

Overtorment commented on screen/settings/electrumSettings.js:
```diff
@@ -164,54 +347,79 @@ export default class ElectrumSettings extends Component {
-                underlineColorAndroid="transparent"
-              />
+            <View style={styles.addServerTitle}>
+              <BlueText style={styles.settingText}>Add custom server</BlueText>
```
add to localization, and other strings

Overtorment reviewed (changes requested):
whole concept feels to be overenginered.
it can be greately simplified if you treat electrums list more like a history list:
that is, logic with `AppStorage.ELECTRUM_HOST` stays and doesnt get replaced with `AppStorage.ELECTRUM_SERVER_LIST`. every time user inputs a working unique electrum - it gets pushed to history list, which should be rendered as a list the way it is rendered now. 
"Add custom server" & "Use SSL" are unnecessary

PS. sorry it took so long to review. Lets finish it up and move to more testing

misovan commented:
OK, thanks for review. I'll try to get back to that in few days. 

1. I'll remove "selected" property from the list and adapt related logic to use previous ELECTRUM_HOST, etc constants from local storage. 
2. I'll revamp the UI so the list is presented as a history. 
3. I'll think of whether to keep "Remove" option alongside server item or perhaps replace it by "Remove all" that would clean the UI from abundance of buttons.
4. After this new concept gets ack, I'll deal with localization and other necessary cleaning

misovan commented:
Ok, I have updated the pullrequest so it now implements more simple electrum server list history with 3 features: 
- new server is added to list on save
- server from list can be selected and becomes active server (after restart)
- history can be cleaned

Open topics:
- Translation and some small styling after ack the concept by reviewer
- Open whether to keep existing feature to get back to default servers through submitting empty form or make it rather special button

marcosrdz commented:
DeepScan error

marcosrdz commented:
I think Clear All should have a confirmation alert because you might accidentally press it and not have the sever information readily available to reinsert.

marcosrdz commented:
Perhaps a "Reset to default" button to have it connect back at a random server?

marcosrdz reviewed (changes requested):
Deep scan error

misovan commented:
> I think Clear All should have a confirmation alert because you might accidentally press it and not have the sever information readily available to reinsert.

I'll do that.

> Perhaps a "Reset to default" button to have it connect back at a random server?

I'll do that. I'll then refactor save method and add basic validation to form fields.

Overtorment commented on class/app-storage.js:
```diff
@@ -677,4 +678,4 @@ export class AppStorage {
     return new Promise(resolve => setTimeout(resolve, ms));
   };
-}
+}
```
unnecessary change

Overtorment commented on .gitignore:
```diff
@@ -65,6 +65,6 @@ artifacts/
-*.realm
-*.realm.lock
\ No newline at end of file
+/.editorconfig
```
looks like a result of unfortunate conflict resolve. 
lines

```
*.mx
*.realm
*.realm.lock 
```

got lost

Overtorment commented on loc/en.json:
```diff
@@ -223,78 +223,85 @@
-    "saved": "Saved",
-    "success_transaction_broadcasted" : "Success! You transaction has been broadcasted!"
-  },
+    "settings": {
```
only changed (or added) lines should be marked. in this case I sense `\n\r` issue, but not sure

Overtorment commented on screen/settings/electrumSettings.js:
```diff
@@ -66,42 +65,127 @@ export default class ElectrumSettings extends Component {
   };
 
+  selectServer = async server => {
+    const { host, port, sslPort } = server;
```
I think you can throw away body of that function and replace it with

```
this.setState({ host: server.host, port: server.port, sslPort: server.sslPort }, () => {
  this.save();
});

```

Overtorment commented on screen/settings/electrumSettings.js:
```diff
@@ -66,42 +65,127 @@ export default class ElectrumSettings extends Component {
+    this.setState({ isLoading: true }, async () => {
+      await AsyncStorage.setItem(AppStorage.ELECTRUM_SERVER_HISTORY, JSON.stringify([]));
+      try {
+        await DefaultPreference.setName('group.io.bluewallet.bluewallet');
```
you dont need to send ELECTRUM_SERVER_HISTORY to DefaultPreference.
Its only needed to let IOS Widget know which electrum we are using so widget can fetch fee estimations etc

Overtorment commented on screen/settings/electrumSettings.js:
```diff
@@ -66,42 +65,127 @@ export default class ElectrumSettings extends Component {
+            });
+
+            await AsyncStorage.setItem(AppStorage.ELECTRUM_SERVER_HISTORY, JSON.stringify(serverHistory));
+            await DefaultPreference.set(AppStorage.ELECTRUM_SERVER_HISTORY, JSON.stringify(serverHistory));
```
dont send ELECTRUM_SERVER_HISTORY to ios widget

Overtorment commented on screen/settings/electrumSettings.js:
```diff
@@ -288,4 +400,20 @@ const styles = StyleSheet.create({
+    borderBottomColor: BlueCurrentTheme.colors.formBorder,
+    borderBottomWidth: 1,
+  },
 });
```
file `scripts/maccatalystpatches/applypatchesformaccatalyst.sh` should not be changed

Overtorment reviewed (changes requested):
I like what I see, probably another round of fixes and we are close to merge!

Overtorment commented:
and please do a git rebase so we are not carrying history of 21 commits here

misovan commented:
- I incorporated the comments
- In case of multiple async-storage operations, they are now enclosed into single promise
- I made my best to rebase and simplify commit history to 3 commits (hopefully that one is ok)

Overtorment commented:
I reworked it into https://github.com/BlueWallet/BlueWallet/pull/2423

misovan commented:
> I reworked it into #2423

Thanks, I had cleaned issues with code marked as changed perhaps due to LF/CRLF in translation file however after pushing it, github showed it again as changed... may be something related to my Visual Studio on windows setup, I'll investigate.  To clean up the mess I as well planned  to move the code to clean branch, but did not make it during last couple of days.

I am sorry for unwanted additional work and I wish you & all the team peaceful Christmas.

Overtorment commented:
Happy Xmas!
